File: suites/osprey-compute-kernels/reference/cases/coprime/coprime.hs
-- Count coprime pairs (i,j), 1<=i,j<=N — nested iteration + Euclidean gcd.
gcd' :: Int -> Int -> Int
gcd' a b = if b == 0 then a else gcd' b (a `mod` b)

loopJ :: Int -> Int -> Int -> Int
loopJ _ 0 acc = acc
loopJ i j acc = loopJ i (j - 1) (if gcd' i j == 1 then acc + 1 else acc)

loopI :: Int -> Int -> Int -> Int
loopI 0 _ acc = acc
loopI i n acc = loopI (i - 1) n (loopJ i n acc)

main :: IO ()
main = print (loopI 2000 2000 0)

File: suites/osprey-compute-kernels/reference/cases/coprime/coprime.rs
// Count coprime pairs (i,j), 1<=i,j<=N — nested iteration + Euclidean gcd.
fn gcd(a: i64, b: i64) -> i64 {
    if b == 0 { a } else { gcd(b, a % b) }
}

fn main() {
    let n: i64 = 2000;
    let mut acc: i64 = 0;
    let mut i: i64 = n;
    while i > 0 {
        let mut j: i64 = n;
        while j > 0 {
            if gcd(i, j) == 1 {
                acc += 1;
            }
            j -= 1;
        }
        i -= 1;
    }
    println!("{}", acc);
}
